quaireaux/math/src/sha256.cairo

```diff
@@ -0,0 +1,321 @@
+
+fn pow(x: u128, n: u128) -> u128 {
+    quaireaux_utils::check_gas();
+    if n == 0 {
+        1
+    } else if (n & 1) == 1 {
+        x * pow(x * x, n / 2)
+    } else {
+        pow(x * x, n / 2)
+    }
+}
+
+fn shl(x: u128, n: u128) -> u128 {
+    x * pow(2, n)
+}
+
+fn shr(x: u128, n: u128) -> u128 {
+    x / pow(2, n)
+}
+
+fn ch(x: u32, y: u32, z: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let y: u128 = upcast(y);
+    let z: u128 = upcast(z);
+    let result = (x & y) ^ ((x ^ U32_MAX) & z);
+    downcast(result).unwrap()
+}
+
+fn maj(x: u32, y: u32, z: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let y: u128 = upcast(y);
+    let z: u128 = upcast(z);
+    let result = (x & y) ^ (x & z) ^ (y & z);
+    downcast(result).unwrap()
+}
+
+fn bsig0(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 2) | shl(x, 32 - 2)) & U32_MAX;
+    let x2 = (shr(x, 13) | shl(x, 32 - 13)) & U32_MAX;
+    let x3 = (shr(x, 22) | shl(x, 32 - 22)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn bsig1(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 6) | shl(x, 32 - 6)) & U32_MAX;
+    let x2 = (shr(x, 11) | shl(x, 32 - 11)) & U32_MAX;
+    let x3 = (shr(x, 25) | shl(x, 32 - 25)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn ssig0(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 7) | shl(x, 32 - 7)) & U32_MAX;
+    let x2 = (shr(x, 18) | shl(x, 32 - 18)) & U32_MAX;
+    let x3 = (shr(x, 3)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn ssig1(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 17) | shl(x, 32 - 17)) & U32_MAX;
+    let x2 = (shr(x, 19) | shl(x, 32 - 19)) & U32_MAX;
+    let x3 = (shr(x, 10)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn sha256(mut data: Array<u8>) -> Array<u8> {
+    let u64_data_length: felt252 = (data.len() * 8).into();
+    let u64_data_length: felt252 = (data.len() * 8).into();
+    let u64_data_length: u128 = u64_data_length.try_into().unwrap();
+    let u64_data_length = u64_data_length & U64_MAX;
+
+    // add one
+    data.append(0x80);
+
+    add_padding(ref data);
+
+    // add length to the end
+    data.append(downcast(shr(u64_data_length, 56) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 48) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 40) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 32) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 24) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 16) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 8) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 0) & U8_MAX).unwrap());
+
+    let u32_data_length = 16 * ((data.len() - 1) / 64 + 1);
+    let mut data = from_u8Array_to_u32Array(ref data, u32_data_length);
```

Comment:
No need to pass as ref here (you'll need to also update `from_u8Array_to_u32Array`)
```suggestion
    let mut data = from_u8Array_to_u32Array(data, u32_data_length);
```